File: problem45/Solution.py
def minReorder(n, connections) -> int:
    edges = { (a,b) for a,b in connections}
    print(f"edges:\n{edges}")
    neighbours = {city:[] for city in range(n)}
    print(f"neighbours:\n{neighbours}")
    visit = set()
    changes = 0
    

    for a,b in connections:
        neighbours[a].append(b)
        neighbours[b].append(a)
    print(f"neighbours:\n{neighbours}")

    def dfs(city):
        nonlocal edges, neighbours, visit, changes

        for neighbour in neighbours[city]:
            if neighbour in visit:
                continue
            if (neighbour, city) not in edges:
                changes += 1
            visit.add(neighbour)
            dfs(neighbour)

    visit.add(0)
    dfs(0)

    return changes

def main():
    connections = [[4,2],[2,3],[1,0],[1,2]]
    print("reorders = ",minReorder(5, connections))
if __name__ == "__main__":
    main()

File: problem45/Solution.java
import java.util.List;
import java.util.ArrayList;

public class Solution {
    private int reorderCount = 0;
    private List<Boolean> visited = new ArrayList<>();
	private List<List<Integer>> graph = new ArrayList<>();

    public static void main(String[] args) {
        int connections[][] = {{1,0},{1,3},{2,3},{4,0},{4,5}};
        System.out.println("Reorders = " + new Solution().minReorder(6, connections));
    }

    public int minReorder(int n, int[][] connections) {
		for (int i = 0; i < n; i++) {
            visited.add(i, false);
            graph.add(i, new ArrayList<>());
        }

		for (int[] pair : connections) {
			graph.get(pair[0]).add(pair[1]);
			graph.get(pair[1]).add(-pair[0]);
		}
		dfs(0);
		return reorderCount;
    }

    private void dfs(int city) {
		if (visited.get(city)) return;
		visited.set(city, true);
		for (int neighbour : graph.get(city)) {
			if (visited.get(Math.abs(neighbour))) continue;
			if (neighbour > 0) reorderCount++;
			dfs(Math.abs(neighbour));
		}
    }
}

